File: examples/Algorithm/python/01_complexity.py
"""
Time Complexity Comparison Experiment
Time Complexity Comparison

Compares the execution times of algorithms with various time complexities.
"""

import time
import random


def measure_time(func, *args):
    """Measure function execution time"""
    start = time.perf_counter()
    result = func(*args)
    end = time.perf_counter()
    return end - start, result


# =============================================================================
# O(1) - Constant Time
# =============================================================================
def constant_time(arr):
    """Return the first element of an array - O(1)"""
    if arr:
        return arr[0]
    return None


# =============================================================================
# O(log n) - Logarithmic Time
# =============================================================================
def binary_search(arr, target):
    """Binary Search - O(log n)"""
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1


# =============================================================================
# O(n) - Linear Time
# =============================================================================
def linear_search(arr, target):
    """Linear Search - O(n)"""
    for i, val in enumerate(arr):
        if val == target:
            return i
    return -1


def find_max(arr):
    """Find Maximum Value - O(n)"""
    if not arr:
        return None
    max_val = arr[0]
    for val in arr:
        if val > max_val:
            max_val = val
    return max_val


# =============================================================================
# O(n log n) - Linearithmic Time
# =============================================================================
def merge_sort(arr):
    """Merge Sort - O(n log n)"""
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result


# =============================================================================
# O(n^2) - Quadratic Time
# =============================================================================
def bubble_sort(arr):
    """Bubble Sort - O(n^2)"""
    arr = arr.copy()
    n = len(arr)
    for i in range(n):
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    return arr


def has_duplicate_naive(arr):
    """Duplicate Check (naive) - O(n^2)"""
    n = len(arr)
    for i in range(n):
        for j in range(i + 1, n):
            if arr[i] == arr[j]:
                return True
    return False


# =============================================================================
# O(2^n) - Exponential Time
# =============================================================================
def fibonacci_recursive(n):
    """Fibonacci (Recursive) - O(2^n)"""
    if n <= 1:
        return n
    return fibonacci_recursive(n - 1) + fibonacci_recursive(n - 2)


def fibonacci_dp(n):
    """Fibonacci (DP) - O(n)"""
    if n <= 1:
        return n
    dp = [0] * (n + 1)
    dp[1] = 1
    for i in range(2, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]
    return dp[n]


# =============================================================================
# Experiment and Results Output
# =============================================================================
def run_experiments():
    """Run time complexity experiments"""
    print("=" * 60)
    print("Time Complexity Comparison Experiment")
    print("=" * 60)

    # Prepare data
    sizes = [100, 1000, 10000]

    for size in sizes:
        print(f"\n[ Array Size: {size} ]")
        print("-" * 40)

        arr = list(range(size))
        random_arr = random.sample(range(size * 10), size)
        target = arr[size // 2]  # Middle value

        # O(1) test
        t, _ = measure_time(constant_time, arr)
        print(f"O(1)     Constant Time:   {t * 1000:.6f} ms")

        # O(log n) test
        t, _ = measure_time(binary_search, arr, target)
        print(f"O(log n) Binary Search:   {t * 1000:.6f} ms")

        # O(n) test
        t, _ = measure_time(linear_search, arr, target)
        print(f"O(n)     Linear Search:   {t * 1000:.6f} ms")

        t, _ = measure_time(find_max, random_arr)
        print(f"O(n)     Find Max:        {t * 1000:.6f} ms")

        # O(n log n) test
        if size <= 10000:
            t, _ = measure_time(merge_sort, random_arr)
            print(f"O(n log n) Merge Sort:  {t * 1000:.6f} ms")

        # O(n^2) test (small sizes only)
        if size <= 1000:
            t, _ = measure_time(bubble_sort, random_arr)
            print(f"O(n^2)   Bubble Sort:     {t * 1000:.6f} ms")

    # O(2^n) vs O(n) Fibonacci comparison
    print("\n[ Fibonacci Comparison: O(2^n) vs O(n) ]")
    print("-" * 40)

    for n in [10, 20, 30]:
        t_recursive, _ = measure_time(fibonacci_recursive, n)
        t_dp, _ = measure_time(fibonacci_dp, n)
        print(f"n={n}: Recursive O(2^n) = {t_recursive * 1000:.4f} ms, "
              f"DP O(n) = {t_dp * 1000:.6f} ms")

    print("\n" + "=" * 60)
    print("Experiment Complete!")
    print("=" * 60)


if __name__ == "__main__":
    run_experiments()

File: examples/Algorithm/c/01_complexity.c
/*
 * Time Complexity
 * Big O Notation and Complexity Analysis
 *
 * A method for analyzing the efficiency of algorithms.
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

/* =============================================================================
 * 1. O(1) - Constant Time
 * ============================================================================= */

int constant_time(int arr[], int n) {
    /* Takes the same amount of time regardless of array size */
    return arr[0];
}

/* =============================================================================
 * 2. O(log n) - Logarithmic Time
 * ============================================================================= */

int binary_search(int arr[], int n, int target) {
    int left = 0, right = n - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (arr[mid] == target)
            return mid;
        else if (arr[mid] < target)
            left = mid + 1;
        else
            right = mid - 1;
    }

    return -1;
}

/* =============================================================================
 * 3. O(n) - Linear Time
 * ============================================================================= */

int linear_search(int arr[], int n, int target) {
    for (int i = 0; i < n; i++) {
        if (arr[i] == target)
            return i;
    }
    return -1;
}

int sum_array(int arr[], int n) {
    int sum = 0;
    for (int i = 0; i < n; i++) {
        sum += arr[i];
    }
    return sum;
}

/* =============================================================================
 * 4. O(n log n) - Linearithmic Time
 * ============================================================================= */

void merge(int arr[], int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    int *L = malloc(n1 * sizeof(int));
    int *R = malloc(n2 * sizeof(int));

    for (int i = 0; i < n1; i++)
        L[i] = arr[left + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[mid + 1 + j];

    int i = 0, j = 0, k = left;
    while (i < n1 && j < n2) {
        if (L[i] <= R[j])
            arr[k++] = L[i++];
        else
            arr[k++] = R[j++];
    }

    while (i < n1)
        arr[k++] = L[i++];
    while (j < n2)
        arr[k++] = R[j++];

    free(L);
    free(R);
}

void merge_sort(int arr[], int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        merge_sort(arr, left, mid);
        merge_sort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

/* =============================================================================
 * 5. O(n^2) - Quadratic Time
 * ============================================================================= */

void bubble_sort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

int count_pairs(int arr[], int n) {
    int count = 0;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            count++;
        }
    }
    return count;  /* n*(n-1)/2 */
}

/* =============================================================================
 * 6. O(2^n) - Exponential Time
 * ============================================================================= */

int fibonacci_recursive(int n) {
    if (n <= 1)
        return n;
    return fibonacci_recursive(n - 1) + fibonacci_recursive(n - 2);
}

/* O(n) optimized version */
int fibonacci_iterative(int n) {
    if (n <= 1)
        return n;

    int prev2 = 0, prev1 = 1;
    for (int i = 2; i <= n; i++) {
        int curr = prev1 + prev2;
        prev2 = prev1;
        prev1 = curr;
    }
    return prev1;
}

/* =============================================================================
 * 7. Time Measurement Utility
 * ============================================================================= */

double measure_time(void (*func)(int*, int), int arr[], int n) {
    clock_t start = clock();
    func(arr, n);
    clock_t end = clock();
    return (double)(end - start) / CLOCKS_PER_SEC;
}

/* =============================================================================
 * 8. Space Complexity Examples
 * ============================================================================= */

/* O(1) space */
void reverse_in_place(int arr[], int n) {
    for (int i = 0; i < n / 2; i++) {
        int temp = arr[i];
        arr[i] = arr[n - 1 - i];
        arr[n - 1 - i] = temp;
    }
}

/* O(n) space */
int* reverse_with_copy(int arr[], int n) {
    int *result = malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        result[i] = arr[n - 1 - i];
    }
    return result;
}

/* =============================================================================
 * Test
 * ============================================================================= */

void print_array(int arr[], int n) {
    printf("    [");
    for (int i = 0; i < n && i < 10; i++) {
        printf("%d", arr[i]);
        if (i < n - 1 && i < 9) printf(", ");
    }
    if (n > 10) printf(", ...");
    printf("]\n");
}

int main(void) {
    printf("============================================================\n");
    printf("Time Complexity Examples\n");
    printf("============================================================\n");

    /* 1. O(1) */
    printf("\n[1] O(1) - Constant Time\n");
    int arr1[] = {5, 2, 8, 1, 9};
    printf("    First element: %d\n", constant_time(arr1, 5));

    /* 2. O(log n) */
    printf("\n[2] O(log n) - Binary Search\n");
    int arr2[] = {1, 3, 5, 7, 9, 11, 13, 15};
    int idx = binary_search(arr2, 8, 7);
    printf("    Array: [1,3,5,7,9,11,13,15]\n");
    printf("    Position of 7: %d\n", idx);

    /* 3. O(n) */
    printf("\n[3] O(n) - Linear Search\n");
    int arr3[] = {4, 2, 7, 1, 9, 3};
    printf("    Array sum: %d\n", sum_array(arr3, 6));

    /* 4. O(n log n) */
    printf("\n[4] O(n log n) - Merge Sort\n");
    int arr4[] = {64, 34, 25, 12, 22, 11, 90};
    printf("    Before sort: ");
    print_array(arr4, 7);
    merge_sort(arr4, 0, 6);
    printf("    After sort: ");
    print_array(arr4, 7);

    /* 5. O(n^2) */
    printf("\n[5] O(n^2) - Bubble Sort\n");
    int arr5[] = {64, 34, 25, 12, 22, 11, 90};
    bubble_sort(arr5, 7);
    printf("    After sort: ");
    print_array(arr5, 7);
    printf("    Number of pairs from 5 elements: %d\n", count_pairs(arr5, 5));

    /* 6. O(2^n) vs O(n) */
    printf("\n[6] O(2^n) vs O(n) - Fibonacci\n");
    printf("    Fibonacci(20) recursive: %d\n", fibonacci_recursive(20));
    printf("    Fibonacci(20) iterative: %d\n", fibonacci_iterative(20));
    printf("    Fibonacci(40) iterative: %d\n", fibonacci_iterative(40));

    /* 7. Space Complexity */
    printf("\n[7] Space Complexity\n");
    int arr7[] = {1, 2, 3, 4, 5};
    printf("    Original: ");
    print_array(arr7, 5);
    reverse_in_place(arr7, 5);
    printf("    O(1) space reverse: ");
    print_array(arr7, 5);

    /* 8. Complexity Summary */
    printf("\n[8] Complexity Summary\n");
    printf("    | Complexity | 1000 ops    | Example           |\n");
    printf("    |------------|-------------|-------------------|\n");
    printf("    | O(1)       | 1           | Array indexing    |\n");
    printf("    | O(log n)   | 10          | Binary search     |\n");
    printf("    | O(n)       | 1000        | Linear search     |\n");
    printf("    | O(n log n) | 10000       | Merge sort        |\n");
    printf("    | O(n^2)     | 1000000     | Bubble sort       |\n");
    printf("    | O(2^n)     | Very large  | All subsets       |\n");

    printf("\n============================================================\n");

    return 0;
}
